File: go/native/api_test.go
// Copyright The KCL Authors. All rights reserved.
//
// This file contains the Go test cases corresponding to the Python KCL API tests.

package native

import (
	"os"
	"strings"
	"testing"

	"kcl-lang.io/lib/go/api"
)

const (
	testFileSchema        = "./../test_data/schema.k"
	testFileOptionMain    = "./../test_data/option/main.k"
	testFileOverrideBak   = "./../test_data/override_file/main.bak"
	testFileOverrideMain  = "./../test_data/override_file/main.k"
	testFileFormatPath    = "./../test_data/format_path/test.k"
	testFileLintPath      = "./../test_data/lint_path/test-lint.k"
	testFileRenameMain    = "./../test_data/rename/main.k"
	testFileRenameBak     = "./../test_data/rename/main.bak"
	testFileTestingPkg    = "./../test_data/testing/..."
	testFileSettingsYaml  = "./../test_data/settings/kcl.yaml"
	testFileUpdateDep     = "./../test_data/update_dependencies"
	testFileUpdateDepMain = "./../test_data/update_dependencies/main.k"
	testWorkDir           = "./../test_data"
)

func TestExecAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.ExecProgramArgs{
		KFilenameList: []string{testFileSchema},
	}

	result, err := client.ExecProgram(args)
	if err != nil {
		t.Fatalf("ExecProgram failed: %v", err)
	}

	expectedYaml := "app:\n  replicas: 2"
	if result.YamlResult != expectedYaml {
		t.Errorf("Expected YAML:\n%s\nGot:\n%s", expectedYaml, result.YamlResult)
	}
}

func TestExecAPIFailed(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.ExecProgramArgs{
		KFilenameList: []string{"file_not_found"},
	}

	_, err := client.ExecProgram(args)
	if err == nil {
		t.Fatal("Expected error for non-existent file, but got nil")
	}

	if !strings.Contains(err.Error(), "Cannot find the kcl file") {
		t.Errorf("Expected error to contain 'Cannot find the kcl file', got: %v", err)
	}
}

func TestParseProgramAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.ParseProgramArgs{
		Paths: []string{testFileSchema},
	}

	result, err := client.ParseProgram(args)
	if err != nil {
		t.Fatalf("ParseProgram failed: %v", err)
	}

	if len(result.Paths) != 1 {
		t.Errorf("Expected 1 path, got %d", len(result.Paths))
	}
	if len(result.Errors) != 0 {
		t.Errorf("Expected 0 parse errors, got %d", len(result.Errors))
	}
}

func TestParseFileAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.ParseFileArgs{
		Path: testFileSchema,
	}

	result, err := client.ParseFile(args)
	if err != nil {
		t.Fatalf("ParseFile failed: %v", err)
	}

	if len(result.Deps) != 0 {
		t.Errorf("Expected 0 dependencies, got %d", len(result.Deps))
	}
	if len(result.Errors) != 0 {
		t.Errorf("Expected 0 parse errors, got %d", len(result.Errors))
	}
}

func TestLoadPackageAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.LoadPackageArgs{
		ParseArgs: &api.ParseProgramArgs{
			Paths: []string{testFileSchema},
		},
		ResolveAst: true,
	}

	result, err := client.LoadPackage(args)
	if err != nil {
		t.Fatalf("LoadPackage failed: %v", err)
	}

	found := false
	for _, sym := range result.Symbols {
		if sym.Ty != nil && sym.Ty.SchemaName == "AppConfig" {
			found = true
			break
		}
	}

	if !found {
		t.Error("Expected to find symbol with schema name 'AppConfig'")
	}
}

func TestListVariablesAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.ListVariablesArgs{
		Files: []string{testFileSchema},
	}

	result, err := client.ListVariables(args)
	if err != nil {
		t.Fatalf("ListVariables failed: %v", err)
	}

	appVar, ok := result.Variables["app"]
	if !ok {
		t.Fatal("Expected variable 'app' not found")
	}

	if len(appVar.Variables) == 0 {
		t.Fatal("Expected at least one variable under 'app'")
	}

	expectedValue := "AppConfig {\n    replicas: 2\n}"
	if appVar.Variables[0].Value != expectedValue {
		t.Errorf("Expected value:\n%s\nGot:\n%s", expectedValue, appVar.Variables[0].Value)
	}
}

func TestListOptionsAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.ParseProgramArgs{
		Paths: []string{testFileOptionMain},
	}

	result, err := client.ListOptions(args)
	if err != nil {
		t.Fatalf("ListOptions failed: %v", err)
	}

	if len(result.Options) != 3 {
		t.Errorf("Expected 3 options, got %d", len(result.Options))
	}

	expectedNames := []string{"key1", "key2", "metadata-key"}
	for i, name := range expectedNames {
		if result.Options[i].Name != name {
			t.Errorf("Option %d: expected '%s', got '%s'", i, name, result.Options[i].Name)
		}
	}
}

func TestGetSchemaTypeAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.GetSchemaTypeMappingArgs{
		ExecArgs: &api.ExecProgramArgs{
			KFilenameList: []string{testFileSchema},
		},
	}

	result, err := client.GetSchemaTypeMapping(args)
	if err != nil {
		t.Fatalf("GetSchemaTypeMapping failed: %v", err)
	}

	appSchema, ok := result.SchemaTypeMapping["app"]
	if !ok {
		t.Fatal("Expected schema 'app' not found")
	}

	checkPropType := func(propName, expectedType string) {
		prop, ok := appSchema.Properties[propName]
		if !ok {
			t.Fatalf("Property '%s' not found in app schema", propName)
		}
		if prop.Type != expectedType {
			t.Errorf("Property '%s': expected type '%s', got '%s'", propName, expectedType, prop.Type)
		}
	}

	checkPropType("replicas", "int")
	checkPropType("my_func", "function")
	checkPropType("maps", "schema")

	mapsProp := appSchema.Properties["maps"]
	if mapsProp.IndexSignature == nil {
		t.Fatal("Expected index signature for 'maps'")
	}

	if *mapsProp.IndexSignature.KeyName != "name" {
		t.Errorf("Expected key name 'name', got '%s'", *mapsProp.IndexSignature.KeyName)
	}
	if mapsProp.IndexSignature.Key.Type != "str" {
		t.Errorf("Expected key type 'str', got '%s'", mapsProp.IndexSignature.Key.Type)
	}
	if mapsProp.IndexSignature.Val.Type != "schema" {
		t.Errorf("Expected value type 'schema', got '%s'", mapsProp.IndexSignature.Val.Type)
	}

	nameProp := mapsProp.IndexSignature.Val.Properties["name"]
	if nameProp.Type != "str" {
		t.Errorf("Expected 'name' type 'str', got '%s'", nameProp.Type)
	}
}

func TestOverrideFileAPI(t *testing.T) {
	client := NewNativeServiceClient()

	bakContent, err := os.ReadFile(testFileOverrideBak)
	if err != nil {
		t.Fatalf("Failed to read bak file: %v", err)
	}
	if err := os.WriteFile(testFileOverrideMain, bakContent, 0644); err != nil {
		t.Fatalf("Failed to write test file: %v", err)
	}

	args := &api.OverrideFileArgs{
		File:  testFileOverrideMain,
		Specs: []string{"b.a=2"},
	}

	result, err := client.OverrideFile(args)
	if err != nil {
		t.Fatalf("OverrideFile failed: %v", err)
	}

	if len(result.ParseErrors) != 0 {
		t.Errorf("Expected 0 parse errors, got %d", len(result.ParseErrors))
	}
	if !result.Result {
		t.Error("Expected override result to be true")
	}

	expectedContent := `a = 1
b = {
    "a": 2
    "b": 2
}
`
	content, err := os.ReadFile(testFileOverrideMain)
	if err != nil {
		t.Fatalf("Failed to read overridden file: %v", err)
	}

	if string(content) != expectedContent {
		t.Errorf("Expected content:\n%s\nGot:\n%s", expectedContent, string(content))
	}
}

func TestFormatCodeAPI(t *testing.T) {
	client := NewNativeServiceClient()

	sourceCode := `schema Person:
    name:   str
    age:    int

    check:
        0 <   age <   120
`
	args := &api.FormatCodeArgs{
		Source: sourceCode,
	}

	result, err := client.FormatCode(args)
	if err != nil {
		t.Fatalf("FormatCode failed: %v", err)
	}

	expectedFormatted := `schema Person:
    name: str
    age: int

    check:
        0 < age < 120

`
	if string(result.Formatted) != expectedFormatted {
		t.Errorf("Expected formatted code:\n%s\nGot:\n%s", expectedFormatted, string(result.Formatted))
	}
}

func TestLintPathAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.LintPathArgs{
		Paths: []string{testFileLintPath},
	}

	result, err := client.LintPath(args)
	if err != nil {
		t.Fatalf("LintPath failed: %v", err)
	}

	if !strings.Contains(result.Results[0], "Module 'math' imported but unused") {
		t.Errorf("Expected lint result to contain 'Module 'math' imported but unused', got: %s", result.Results)
	}
}

func TestValidateCodeAPI(t *testing.T) {
	client := NewNativeServiceClient()

	code := `schema Person:
    name: str
    age: int

    check:
        0 < age < 120
`
	data := `{"name": "Alice", "age": 10}`

	args := &api.ValidateCodeArgs{
		Code:   code,
		Data:   data,
		Format: "json",
	}

	result, err := client.ValidateCode(args)
	if err != nil {
		t.Fatalf("ValidateCode failed: %v", err)
	}

	if !result.Success {
		t.Error("Expected validation to succeed")
	}
	if result.ErrMessage != "" {
		t.Errorf("Expected empty error message, got: %s", result.ErrMessage)
	}
}

func TestRenameAPI(t *testing.T) {
	client := NewNativeServiceClient()

	bakContent, err := os.ReadFile(testFileRenameBak)
	if err != nil {
		t.Fatalf("Failed to read rename bak file: %v", err)
	}
	if err := os.WriteFile(testFileRenameMain, bakContent, 0644); err != nil {
		t.Fatalf("Failed to write rename test file: %v", err)
	}

	args := &api.RenameArgs{
		PackageRoot: "./../test_data/rename",
		SymbolPath:  "a",
		FilePaths:   []string{testFileRenameMain},
		NewName:     "a2",
	}

	_, err = client.Rename(args)
	if err != nil {
		t.Fatalf("Rename failed: %v", err)
	}
}

func TestRenameCodeAPI(t *testing.T) {
	client := NewNativeServiceClient()

	args := &api.RenameCodeArgs{
		PackageRoot: "/mock/path",
		SymbolPath:  "a",
		SourceCodes: map[string]string{
			"/mock/path/main.k": "a = 1\nb = a",
		},
		NewName: "a2",
	}

	result, err := client.RenameCode(args)
	if err != nil {
		t.Fatalf("RenameCode failed: %v", err)
	}

	expectedCode := "a2 = 1\nb = a2"
	actualCode, ok := result.ChangedCodes["/mock/path/main.k"]
	if !ok {
		t.Fatal("Expected changed code for '/mock/path/main.k'")
	}

	if actualCode != expectedCode {
		t.Errorf("Expected code:\n%s\nGot:\n%s", expectedCode, actualCode)
	}
}

func TestTestingAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.TestArgs{
		PkgList: []string{testFileTestingPkg},
	}

	result, err := client.Test(args)
	if err != nil {
		t.Fatalf("Test failed: %v", err)
	}

	if len(result.Info) != 2 {
		t.Errorf("Expected 2 info entries, got %d", len(result.Info))
	}
}

func TestLoadSettingsFilesAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.LoadSettingsFilesArgs{
		WorkDir: testWorkDir,
		Files:   []string{testFileSettingsYaml},
	}

	result, err := client.LoadSettingsFiles(args)
	if err != nil {
		t.Fatalf("LoadSettingsFiles failed: %v", err)
	}

	// 验证配置
	if len(result.KclCliConfigs.Files) != 0 {
		t.Errorf("Expected 0 files in KclCliConfigs, got %d", len(result.KclCliConfigs.Files))
	}
	if !result.KclCliConfigs.StrictRangeCheck {
		t.Error("Expected StrictRangeCheck to be true")
	}

	if len(result.KclOptions) != 1 {
		t.Errorf("Expected 1 KclOption, got %d", len(result.KclOptions))
	}
	opt := result.KclOptions[0]
	if opt.Key != "key" || opt.Value != `"value"` {
		t.Errorf("Expected option key 'key' and value '\"value\"', got key '%s' and value '%s'", opt.Key, opt.Value)
	}
}

func TestUpdateDependenciesAPI(t *testing.T) {
	client := NewNativeServiceClient()
	args := &api.UpdateDependenciesArgs{
		ManifestPath: testFileUpdateDep,
	}

	result, err := client.UpdateDependencies(args)
	if err != nil {
		t.Fatalf("UpdateDependencies failed: %v", err)
	}

	pkgNames := make([]string, 0, len(result.ExternalPkgs))
	for _, pkg := range result.ExternalPkgs {
		pkgNames = append(pkgNames, pkg.PkgName)
	}

	if len(pkgNames) != 2 {
		t.Errorf("Expected 2 external packages, got %d", len(pkgNames))
	}
	if !strings.Contains(strings.Join(pkgNames, ","), "helloworld") {
		t.Error("Expected package 'helloworld' in external packages")
	}
	if !strings.Contains(strings.Join(pkgNames, ","), "flask") {
		t.Error("Expected package 'flask' in external packages")
	}
}

func TestExecAPIWithExternalDependencies(t *testing.T) {
	client := NewNativeServiceClient()

	updateArgs := &api.UpdateDependenciesArgs{
		ManifestPath: testFileUpdateDep,
	}
	updateResult, err := client.UpdateDependencies(updateArgs)
	if err != nil {
		t.Fatalf("UpdateDependencies failed: %v", err)
	}

	execArgs := &api.ExecProgramArgs{
		KFilenameList: []string{testFileUpdateDepMain},
		ExternalPkgs:  updateResult.ExternalPkgs,
	}
	execResult, err := client.ExecProgram(execArgs)
	if err != nil {
		t.Fatalf("ExecProgram failed: %v", err)
	}

	expectedYaml := "a: Hello World!"
	if execResult.YamlResult != expectedYaml {
		t.Errorf("Expected YAML:\n%s\nGot:\n%s", expectedYaml, execResult.YamlResult)
	}
}

func TestGetVersionAPI(t *testing.T) {
	client := NewNativeServiceClient()

	result, err := client.GetVersion(&api.GetVersionArgs{})
	if err != nil {
		t.Fatalf("GetVersion failed: %v", err)
	}

	resultStr := result.String()
	if !strings.Contains(resultStr, "Version") {
		t.Error("Expected version info to contain 'Version'")
	}
	if !strings.Contains(resultStr, "GitCommit") {
		t.Error("Expected version info to contain 'GitCommit'")
	}
}

File: wasm/tests/api.test.ts
import { expect, test } from "@jest/globals";
import { init, MemFS } from "@wasmer/wasi";
import * as kcl from "../src";

async function loadWithMemFS(
  files: Record<string, string> = {}
): Promise<{ inst: WebAssembly.Instance; fs: MemFS }> {
  await init();
  const fs = new MemFS();
  for (const [path, content] of Object.entries(files)) {
    const dir = path.slice(0, path.lastIndexOf("/"));
    if (dir && dir !== "/") {
      try {
        fs.createDir(dir);
      } catch {
        // directory already exists
      }
    }
    const f = fs.open(path, { read: true, write: true, create: true });
    f.writeString(content);
    f.free();
  }
  const inst = await kcl.load({ fs });
  return { inst, fs };
}

test("ping echoes the value", async () => {
  const inst = await kcl.load();
  const result = kcl.ping(inst, { value: "hello" });
  expect(result.value).toBe("hello");
});

test("getVersion returns version information", async () => {
  const inst = await kcl.load();
  const result = kcl.getVersion(inst);
  expect(result.version).toMatch(/^\d+\.\d+\.\d+/);
  expect(result.checksum).not.toBe("");
  expect(result.gitSha).not.toBe("");
  expect(result.versionInfo).toContain("Version:");
});

test("parseProgram parses inline sources", async () => {
  const inst = await kcl.load();
  const result = kcl.parseProgram(inst, { sources: ["a = 1\nb = a"] });
  expect(result.errors).toEqual([]);
  expect(JSON.parse(result.astJson)).toBeTruthy();
});

test("parseFile parses a single file source", async () => {
  const inst = await kcl.load();
  const result = kcl.parseFile(inst, {
    path: "/mock/main.k",
    source: "schema Person:\n  name: str\n",
  });
  expect(result.errors).toEqual([]);
  expect(JSON.parse(result.astJson)).toBeTruthy();
});

test("listOptions lists option help from sources", async () => {
  const inst = await kcl.load();
  const result = kcl.listOptions(inst, {
    sources: [
      'option1 = option("option1", type="str", required=True, default="", help="option 1 help")\n',
    ],
  });
  expect(result.options).toHaveLength(1);
  expect(result.options[0].name).toBe("option1");
  expect(result.options[0].type).toBe("str");
  expect(result.options[0].required).toBe(true);
  expect(result.options[0].help).toBe("option 1 help");
});

test("lintPath reports unused imports from the sandbox filesystem", async () => {
  const { inst } = await loadWithMemFS({ "/test.k": "import math\n\na = 1\n" });
  const result = kcl.lintPath(inst, { paths: ["/test.k"] });
  expect(result.results).toEqual(["Module 'math' imported but unused"]);
});

test("lintPath reports missing files in the results", async () => {
  const inst = await kcl.load();
  const result = kcl.lintPath(inst, { paths: ["/no/such/file.k"] });
  expect(result.results[0]).toContain("Cannot find the kcl file");
});

test("renameCode returns the changed code", async () => {
  const inst = await kcl.load();
  const result = kcl.renameCode(inst, {
    packageRoot: "/mock/path",
    symbolPath: "a",
    sourceCodes: { "/mock/path/main.k": "a = 1\nb = a" },
    newName: "a2",
  });
  expect(result.changedCodes["/mock/path/main.k"]).toBe("a2 = 1\nb = a2");
});

test("getSchemaTypeMapping returns schema types from code", async () => {
  const inst = await kcl.load();
  const result = kcl.getSchemaTypeMapping(inst, {
    execArgs: { kCodeList: ["schema Person:\n  name: str\n  age: int = 18"] },
    schemaName: "Person",
  });
  const person = result.schemaTypeMapping["Person"];
  expect(person.type).toBe("schema");
  expect(person.schemaName).toBe("Person");
  expect(person.properties["name"].type).toBe("str");
  expect(person.properties["age"].default).toBe("18");
  expect(person.required).toEqual(["name", "age"]);
});

test("formatPath formats files in the sandbox filesystem", async () => {
  const { inst } = await loadWithMemFS({ "/fmt.k": "a   =   1\n" });
  const result = kcl.formatPath(inst, { path: "/fmt.k" });
  expect(result.changedPaths).toEqual(["/fmt.k"]);
});

test("overrideFile overrides files in the sandbox filesystem", async () => {
  const { inst } = await loadWithMemFS({ "/ovr.k": "a = 1\n" });
  const result = kcl.overrideFile(inst, {
    file: "/ovr.k",
    specs: ["a=2"],
  });
  expect(result.result).toBe(true);
  expect(result.parseErrors).toEqual([]);
});

test("loadSettingsFiles loads settings from the sandbox filesystem", async () => {
  const { inst } = await loadWithMemFS({
    "/kcl.yaml":
      "kcl_cli_configs:\n  strict_range_check: true\nkcl_options:\n  - key: app-name\n    value: kcl\n",
  });
  const result = kcl.loadSettingsFiles(inst, {
    workDir: "/",
    files: ["/kcl.yaml"],
  });
  expect(result.kclCliConfigs.strictRangeCheck).toBe(true);
  expect(result.kclOptions).toEqual([{ key: "app-name", value: '"kcl"' }]);
});

test("rename rewrites symbols in sandbox files", async () => {
  const { inst } = await loadWithMemFS({ "/main.k": "a = 1\nb = a\n" });
  const result = kcl.rename(inst, {
    packageRoot: "/",
    symbolPath: "a",
    filePaths: ["/main.k"],
    newName: "a2",
  });
  expect(result.changedFiles).toEqual(["/main.k"]);
});

test("loadPackage loads the semantic model", async () => {
  const inst = await kcl.load();
  const result = kcl.loadPackage(inst, {
    parseArgs: { sources: ["schema Person:\n  name: str\n"] },
    resolveAst: true,
  });
  expect(result.parseErrors).toEqual([]);
  expect(result.typeErrors).toEqual([]);
  expect(JSON.parse(result.program)).toBeTruthy();
});

test("listVariables lists variables from the sandbox filesystem", async () => {
  const { inst } = await loadWithMemFS({ "/vars.k": 'a = 1\nb = "s"\n' });
  const result = kcl.listVariables(inst, {
    files: ["/vars.k"],
    specs: ["a", "b"],
  });
  expect(result.parseErrors).toEqual([]);
  expect(result.unsupportedCodes).toEqual([]);
  expect(result.variables["a"][0].value).toBe("1");
  expect(result.variables["a"][0].opSym).toBe("=");
  expect(result.variables["b"][0].value).toBe('"s"');
  expect(result.variables["b"][0].opSym).toBe("=");
});

test("validateCode validates data against a schema", async () => {
  const inst = await kcl.load();
  const result = kcl.validateCode(inst, {
    code: "schema Person:\n  name: str\n  age: int\n  check:\n    0 < age < 120",
    data: '{"name": "Alice", "age": 10}',
  });
  expect(result.success).toBe(true);
});

test("validateCode reports invalid KCL as a graceful result", async () => {
  const inst = await kcl.load();
  const result = kcl.validateCode(inst, {
    code: "schema Person:\n  name: str\n  age: int\n  check: 0 < age < 120",
    data: '{"name": "Alice", "age": 10}',
  });
  expect(result.success).toBe(false);
  expect(result.errMessage).toMatch(/missing expression/);
});

test("updateDependencies reports it is not supported in the WASM build", async () => {
  const { inst } = await loadWithMemFS({
    "/proj/kcl.mod": '[package]\nname = "proj"\nversion = "0.1.0"\n',
  });
  expect(() => kcl.updateDependencies(inst, { manifestPath: "/proj" })).toThrow(
    /not supported in the WASM build/
  );
});
